### ISSUE TITLE
[Feature Request]: Random patches

### ISSUE THREAD

**Issue by buruto** [PCSX2/pcsx2#7523]
### Description

Some PS2 roms, like Kingdom Hearts Re Chain of Memories, reads from the inserted game or rom disc any function to the RAM on a random entry point when it is needed at runtime and on the static code this function is invoked by a `jalr` instruction because the entry point of that function is random and can't be invoked with a `jal` instruction whose operand is always immediate and not variable.

By using memory read and write breakpoints this is possible to find the random entry points of such functions and assemble some of their instructions to achieve some cheats in the game but all these cheats are disabled when these functions are again read from the inserted disc to the RAM on other random addresses and this is helpless to remember the addresses of the instructions that you assembled before because these instructions that you assembled are no longer there but in another random addresses and you need to use memory read and write breakpoints again to find them again and again assemble them one by one.

Once the random entry point of any function has been found, probably by using memory read and write breakpoints, right click on the first instruction of that function, that is probably `addiu sp`, and on the context menu that appears click on the "Create a random patch" option which my feature request is to add that option to that context menu.

Then a dialog appears asking the user to enter a number of instructions to look for  and a string for the name of the random patch that he or she wants to save. Below the textbox for the name of the random patch, there are two assembly views, a "Before" assembly view on the left side of the dialog and a "After" assembly view on the right side of the dialog and between them draw a right arrow from "Before" to "After".

The "Before" assembly view is read only and shows all the selected instructions whose first was right clicked in the main assembly view of the debugger window of PCSX2 and all the instructions that comes after it but how many were copied is determined by the first number entered in the dialog. 

The "After" assembly view also shows what the "Before" assembly view shows but unlike the "Before" assembly view, the "After" assembly view is not read only and the user can assemble all the or some of the instructions and opcodes there.

Every instruction that was assembled in the "After" assembly view will be both highlighted and coloured differently so the user knows which instructions he or she assembled.

The "Before" assembly view can be used to view the original instructions before they were assembled in the "After" assembly view. 

The "After" assembly view is another information which implies all the indexes of the instructions to assemble and what is the new instruction after the assemble.

At last below the "Before" assembly view there is a "Export random patch" button and below the "After" assembly view there is a "Cancel" button.

No need to explain what the "Cancel" button does because the text of the button is self explanatory and this is too obvious.

If the user clicks on the "Export random patch" button then the random patch is saved to a file in the PCSX2 directory or you can show to the user a save file dialog that asks the user where to save the new random patch.

If the user selects a file that does not exist then a new file is created containing the new random patch.

If the user selects a file that does exist then the new random patch is appended to that file and the existing file will contain more than one random patches.

Additionally add a new "Use random patches" button to the debugger window of PCSX2.

The user has to click on this button if he or she wants to use random patches that he or she made.

Clicking on this button shows another dialog of names of random patches that the user made with the PCSX2 debugger in a ListBox or ListView if you want to additionally show the date and time when the random patch was created for example.

All random patches that appear in this ListBox or ListView are loaded from a file in the PCSX2 directory and selected files outside of PCSX2 directory. You can add a "Import random patches" button which can be used to load random patches from files outside of the PCSX2 directory. An open file dialog appears to let the user to select the files.

All what the user needs to do is to highlight all the random patches in the ListBox or ListView that he or she wants to use and at last click on the "Apply" button or "Cancel" if the user regrets or accidentally clicked on the "Use random patches" button.

A random patch consists of 3 details:

1. A name as string to be displayed in a ListBox or ListView.
2. A word array to be used to find the random entry point of the function because every MIPS instruction is a word and a word is always 32 bits in MIPS. Let's call this word array "goal".
3. Another word array of even length where each word at an even index, if the index is zero based, is an index of an instruction in the "goal" array to be assembled, and each word in an odd index is the new instruction. Let's call this word array "patch"

If the random entry point of the function has been found then just add to it the index, which was read at an even index, multiply 4, because a word is always 4 bytes in MIPS,  to calculate the address of the instruction that has to be assembled at runtime and the new instruction was read at an odd index.

As said earlier the "goal" word array is used to find the random entry point of the function by scanning all word arrays whose length is equal to that of the "goal" in the PS2 RAM and compare each scanned word array with the "goal" word array.

The comparison is done by XORing all the bits of the scanned array with the "goal" array and counting all the 0 bits in the result array and this is the 0-bit counter of the scanned array.

The random entry point of the function that the finder seeks is the address of the scanned array whose 0-bit counter is the biggest.

This comparison and search methods can be useful to find functions not only their entry points are random but their code is random too if for example the function contains a `jal` instruction that invokes another function also at random entry point.

This means that every time the operand, the target address, of these `jal` instructions will be different because unlike branch instructions whose target address is always relative, the target address of jumps are always absolute.

If the entry point of any function has changed then all `jal`s that targeted that function must have changed as well.

Because of this not all bits in the result array are 0 because of the `jal` instructions that changed but at least still most of the bits will be 0 if the scanned array is the function that the hacker is looking for.

At last assemble all the instructions of the function, whose random entry point has been found, according to the second word array "patch" in the random patch and that's it.

### Reason

I want to use cheats that modify functions on random entry points. Enabling cheats and using a pnach file is not a solution to this problem because the entry points of these functions are random and thus the addresses of their instructions are random as well.

Without this feature the best way is to use memory read and write breakpoints to find the random entry points of these functions over and over again and then use the "Assemble Opcode" context menu option to change them one by one to enable the cheats.

But with this feature you enable all the cheats by few mouse clicks only and wait a little time after that.

### Examples

In Kingdom Hearts Re Chain of Memories I want to use a cheat that make all card packs in all Moogle Shops free which means that all cards packs in the game do not cost any moogle points at all.

This can be done by finding the conditional branch instruction that tests whether or not Sora has enough moogle points to buy the selected card pack and assemble that instruction to either unconditional branch or nop so that the path that buys the card pack is always taken even if Sora does not have enough Moogle Points to buy the card pack.

Optionally assemble the `sw` instruction that updates the amount of moogle points to a `nop` instruction if the cheater does not desire that the amount of moogle points will be decreased after buying any card pack from the moogle shop.

The address of the moogle points can be found with Cheat Engine and at this address (after turning the most left digit from 2 to 0) add a memory read breakpoint.

The memory read breakpoint should be triggered when the amount of moogle points has to be compared with the cost of the selected card pack that Sora wants to buy.

You have found the `lw` instruction that reads this number from RAM.

All what you have to do is to scroll down a bit to find the conditional branch instruction (probably `beqz`, `bnez`, `beqzl` or `bnezl` after a `slt` or `sltu` instruction) which tests the result of the comparison that was done by the `slt` or `sltu` instruction that was executed before and this is the conditional branch instruction that you want to assemble to either unconditional branch or nop to always take the path that buys the card pack even if Sora does not have enough moogle points to buy it.

After doing this Sora can buy all the card packs from the moogle guy even if he doesn't have enough moogle points to buy even one card pack from him.

But if Sora leaves the current Disney world and continues to the next Disney world and opens another moogle room there then all the instructions that the hacker assembled have been undone, because all the functions of the moogle shop are read from the inserted disc to the RAM at runtime whenever entering a new moogle room.

And worse the entry points of these functions are also different hexadecimal numbers because they are all random so remembering the addresses of the instructions that you assembled is helpless.

Instead of assembling that conditional branch instruction in the main assembly view of the debugger window of PCSX2, scroll up to the random entry point of that function, which probably starts with `addiu sp`, right click the instruction and select "Create a random patch"

Give it a name like "All card packs are free", enter the size of the function by subtracting the end address with the start address of the function.

The start address is already known because you have already right clicked it, that is probably `addiu sp`.

The end address of the function is probably `jr ra` so just scroll down until you see
`jr ra`.

In the "After" assembly view in the dialog that appears after clicking on the "Create a random patch" context menu option assemble that conditional branch instruction and then click on the "Export random patch" button at the end.

At last every time you want all card packs free just click on the "Use random patches" button, import the random patch that you created if it doesn't appear in the ListBox or ListView, then highlight "All card packs are free" and at last click on the "Apply" button.

Great! Now Sora can buy all the card packs from the moogle guy without having enough moogle points even for a single card pack!

After leaving the current moogle room and entering another moogle room if all the card packs are NOT free again then again click on the "Use random patches" button, again highlight "All card packs are free", again click on the "Apply" button and wait a little time.

After that all the card packs are free again and Sora can again buy all of them without having moogle points at all.

Just 3 mouse clicks and a wait and that's it!

Note that there always will be a small delay after clicking on the "Apply" button and before the dialog disappears because during this delay PCSX2 scans the entire RAM of the game or PS2 rom to find the random entry point of the function to assemble it's instructions assuming that the PC that uses this new feature is powerful enough to play modern games that require a high-end PC.

Optionally you can improve the pnach parser of PCSX2 to support random patches so that random patches can also be used in a pnach file if you want to.

After `patch=1,EE,` will come the word array to be used to find the random entry point of the function and the index of the instruction which you want to assemble and at last `,extended,` or `,word,` and the hex code of the new instruction comes here.

You can also add additional export button to export to a pnach file if you want too.

This way random patches can be used **automatically** without having to click any button or do anything manually but better to apply only after the game or PS2 rom read anything from the inserted game or PS2 rom disc to RAM.

Otherwise this may degrade the performance of PCSX2 a lot and render the game unplayable as a result if the entire RAM of the game or PS2 rom is scanned every past moment!

**Comment by stenzek** [comment]
I don't see how this will work for the general case, nor is it really debugger functionality. If you want to try to implement this yourself, go nuts, but I see it as out of scope.

e.g.

> All what you have to do is to scroll down a bit to find the conditional branch instruction (probably beqz, bnez, beqzl or bnezl after a slt or sltu instruction) which tests the result of the comparison that was done by the slt or sltu instruction that was executed before and this is the conditional branch instruction that you want to assemble to either unconditional branch or nop to always take the path that buys the card pack even if Sora does not have enough moogle points to buy it.

Not everything follows this pattern.

> The start address is already known because you have already right clicked it, that is probably addiu sp.

Not all functions start with `addiu sp, nnn`.

> The end address of the function is probably jr ra so just scroll down until you see jr ra.

Not all functions end with `jr ra`. Tail calls as well as `jalr ra, rn`.

> Note that there always will be a small delay after clicking on the "Apply" button and before the dialog disappears because during this delay PCSX2 scans the entire RAM of the game or PS2 rom to find the random entry point of the function to assemble it's instructions assuming that the PC that uses this new feature is powerful enough to play modern games that require a high-end PC.

What about code that is loaded at runtime? What about when a function changed/gets swapped out?

**Comment by buruto** [comment]
> What about code that is loaded at runtime? What about when a function changed/gets swapped out?

The 3 mouse clicks "Use random patches", highlight "All card packs are free" and "Apply" should be done when the moogle shop functions exist in the RAM of course. This means that this should be done when entering a moogle room and not before.

**Comment by stenzek** [comment]
From a brief discussion internally, it seems this is out of scope for the debugger, and like I said, not sure if it's even possible to do in such a generic way. You need to consider that not all games will these patterns.

If anyone wants to take it on, feel free to re-open, but we already have enough work bringing the current functionality over to Qt.

**Comment by buruto** [comment]
This feature is not needed at all for Kingdom Hearts Re Chain of Memories because there is a `lw` instruction that reads all moogle points every time entering a new room.

By assembling it to `li don't remember name of register, 0xffff` Sora always have 65535 moogle points when entering a new room.

So if Sora enters a moogle room then he has 65535 moogle points which is enough to buy all card packs.

Until Sora enters another moogle room his moogle points will be automatically be reset to 65535 so again he will be able to buy all card packs.

You are right. My feature request is unnecessary.

**Comment by stenzek** [comment]
You're still focusing on one game. There's thousands of PS2 games with their own coding style, different compilers, hand-coded ASM, JITted code, etc.

**Comment by buruto** [comment]
> You're still focusing on one game. There's thousands of PS2 games with their own coding style, different compilers, hand-coded ASM, JITted code, etc.

True. Looks like I am an inexperienced user. Maybe I have to do more research before submitting feature requests.

You are good developers and I appreciate your hard work.
